Framework/Source/Graphics/Program/ProgramReflection.cpp

```diff
@@ -0,0 +1,1190 @@
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***************************************************************************/
+#pragma once
+#include "Framework.h"
+#include "ProgramReflection.h"
+#include "Utils/StringUtils.h"
+using namespace slang;
+
+namespace Falcor
+{
+    std::unordered_set<std::string> ProgramReflection::sParameterBlockRegistry;
+    // Represents a "breadcrumb trail" leading from a particular variable
+    // back to the path over member-access and array-indexing operations
+    // that led to it.
+    // E.g., when trying to construct information for `foo.bar[3].baz`
+    // we might have a path that consists of:
+    //
+    // - An entry for the field `baz` in type `Bar` (which knows its offset)
+    // - An entry for element 3 in the type `Bar[]`
+    // - An entry for the field `bar` in type `Foo`
+    // - An entry for the top-level shader parameter `foo`
+    //
+    // To compute the correct offset for `baz` we can walk up this chain
+    // and add up offsets (taking element stride into account for arrays).
+    //
+    // In simple cases, one can track this info top-down, by simply keeping
+    // a "running total" offset, but that doesn't account for the fact that
+    // `baz` might be a texture, UAV, sampler, or uniform, and the offset
+    // we'd need to track for each case is different.
+    struct ReflectionPath
+    {
+        const ReflectionPath* pParent = nullptr;
+        VariableLayoutReflection* pVar = nullptr;
+        TypeLayoutReflection* pTypeLayout = nullptr;
+        uint32_t childIndex = 0;
+    };
+
+    bool isParameterBlockReflection(VariableLayoutReflection* pSlangVar, const std::unordered_set<std::string>& parameterBlockRegistry)
+    {
+        // A candidate for a parameter block must be a top-level constant buffer containing a single struct
+        TypeLayoutReflection* pSlangType = pSlangVar->getTypeLayout();
+        if (pSlangType->getTotalArrayElementCount() == 0 && pSlangType->unwrapArray()->getKind() == TypeReflection::Kind::ConstantBuffer)
+        {
+            if (pSlangType->unwrapArray()->getElementCount() == 1)
+            {
+                TypeLayoutReflection* pFieldLayout = pSlangType->unwrapArray()->getFieldByIndex(0)->getTypeLayout();
+                if (pFieldLayout->getTotalArrayElementCount() == 0 && pFieldLayout->unwrapArray()->getKind() == TypeReflection::Kind::Struct)
+                {
+                    const std::string name = "";
```

Comment:
I assume this is where the Slang support for `struct` type names would plug in?

Alternatively, my hope is that I have enough infrastructure in place that by the end of this week I could have a working `TypeReflection::Kind::ParameterBlock` that you could use to detect this stuff instead...